File: src/main/java/camera/CameraView.java
package camera;

import java.io.File;
import java.util.Optional;
import javafx.geometry.Insets;
import javafx.geometry.Orientation;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Alert;
import javafx.scene.control.Button;
import javafx.scene.control.Separator;
import javafx.scene.control.TextInputDialog;
import javafx.scene.control.TitledPane;
import javafx.scene.image.Image;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Pane;
import javafx.stage.FileChooser;

class CameraView extends TitledPane implements Camera
{
    private static final FileChooser.ExtensionFilter ALL_FILES = new FileChooser.ExtensionFilter("All files", "*.*");

    private static final FileChooser.ExtensionFilter MP4_FILES = new FileChooser.ExtensionFilter("MP4", "*.mp4");

    private static final FileChooser.ExtensionFilter PNG_FILES = new FileChooser.ExtensionFilter("PNG", "*.png");

    private static final FileChooser.ExtensionFilter JPG_FILES = new FileChooser.ExtensionFilter("JPG", "*.jpg");

    private final Configuration configuration;

    private int cameraNum;

    private String uri;

    private final Player player;

    CameraView(Player player, int number, Configuration configuration) {
        this.player = player;
        this.configuration = configuration;
        setCollapsible(false);
        setCamNum(number);

        final BorderPane content = new BorderPane();
        content.getStyleClass().add("video-pane");
        setContent(content);

        final Pane mediaViewPane = new BorderPane(player.getViewNode());
        mediaViewPane.getStyleClass().add("media-view-pane");
        content.setCenter(mediaViewPane);

        content.setBottom(createButtonsPane());

        final String oldURI = configuration.getCameraURI(number);
        if (oldURI != null) {
            try {
                setEncoder(oldURI);
            } catch (Exception e) {
                System.out.println(e.toString());
            }
        }
    }

    private Node createButtonsPane() {
        final Button openFileButton = new Button("Open file");
        openFileButton.setOnAction(event -> openFile());

        final Button openURIButton = new Button("Open URI");
        openURIButton.setOnAction(event -> openURI());

        final Button playButton = new Button("Play");
        playButton.setOnAction(event -> play());

        final Button stopButton = new Button("Stop");
        stopButton.setOnAction(event -> stop());

        final Button screenshotButton = new Button("Screenshot");
        screenshotButton.setOnAction(event -> saveScreenshotToFile());

        final HBox buttonsPanel = new HBox(
                openURIButton,
                openFileButton,
                new Separator(Orientation.VERTICAL),
                playButton,
                stopButton,
                new Separator(Orientation.VERTICAL),
                screenshotButton);
        buttonsPanel.setAlignment(Pos.CENTER_LEFT);
        buttonsPanel.setPadding(new Insets(5.0));
        buttonsPanel.setSpacing(5.0);

        return buttonsPanel;
    }

    private void saveScreenshotToFile() {
        final FileChooser fileChooser = new FileChooser();
        fileChooser.getExtensionFilters().addAll(PNG_FILES, JPG_FILES, ALL_FILES);
        final File file = fileChooser.showSaveDialog(getScene().getWindow());
        if (file == null) {
            return;
        }
        try {
            makeScreenshot(file.toString());
        } catch (Exception e) {
            final Alert alert = new Alert(Alert.AlertType.ERROR);
            alert.setHeaderText("Error saving screenshot");
            alert.setContentText(e.toString());
            alert.showAndWait();
        }
    }

    private void openFile() {
        final FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Open video file");
        fileChooser.getExtensionFilters().addAll(ALL_FILES, MP4_FILES);
        fileChooser.setSelectedExtensionFilter(MP4_FILES);
        final File oldFile = UriUtil.uriToFile(uri);
        if (oldFile != null) {
            fileChooser.setInitialDirectory(oldFile.getParentFile());
            fileChooser.setInitialFileName(oldFile.getName());
        }
        final File file = fileChooser.showOpenDialog(getScene().getWindow());
        if (file != null) {
            try {
                setEncoder(file.toURI().toString());
            } catch (Exception e) {
                showAlert("File open error", e);
            }
        }
    }

    private void openURI() {
        final TextInputDialog dialog = new TextInputDialog(uri);
        dialog.setTitle("Open URI");
        dialog.setHeaderText("Enter streaming URI");
        dialog.setContentText("URI");
        dialog.getDialogPane().setPrefWidth(600.0);
        dialog.setResizable(true);
        final Optional<String> result = dialog.showAndWait();
        try {
            result.ifPresent(this::setEncoder);
        } catch (Exception e) {
            showAlert("URI open error", e);
        }
    }

    private static void showAlert(String headerText, Exception e) {
        final Alert alert = new Alert(Alert.AlertType.ERROR, e.toString());
        alert.setTitle("Error");
        alert.setHeaderText(headerText);
        alert.showAndWait();
    }

    private void stop() {
        player.stop();
    }

    private void play() {
        player.play();
    }

    @Override
    public void setCamNum(int cameraNum) {
        this.cameraNum = cameraNum;
        setText("Camera " + cameraNum);
    }

    @Override
    public int getCamNum() {
        return cameraNum;
    }

    @Override
    public void setEncoder(String uri) {
        player.setSource(uri);
        this.uri = uri;
        configuration.setCameraURI(cameraNum, uri);
    }

    @Override
    public String getEncoder() {
        return uri;
    }

    @Override
    public int getFps() {
        return player.getFps();
    }

    @Override
    public Image getScreenshot() {
        return player.getScreenshot();
    }

    @Override
    public void makeScreenshot(String path) {
        player.makeScreenshot(path);
    }
}

File: src/main/java/camera/CameraView.kt
package camera

import javafx.animation.KeyFrame
import javafx.animation.Timeline
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Orientation
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.control.Alert
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.Separator
import javafx.scene.control.TextInputDialog
import javafx.scene.control.TitledPane
import javafx.scene.image.Image
import javafx.scene.layout.BorderPane
import javafx.scene.layout.HBox
import javafx.scene.layout.Priority
import javafx.scene.layout.Region
import javafx.stage.FileChooser
import javafx.util.Duration

internal class CameraView(
    private val player: Player,
    number: Int,
    private val configuration: Configuration
) : TitledPane(), Camera
{
    override var camNum: Int = 0
        set(cameraNum) {
            field = cameraNum
            text = "Camera $cameraNum"
        }

    override var encoder: String? = null
        set(uri) {
            player.setSource(uri)
            field = uri
            configuration.setCameraURI(camNum, uri)
        }

    override val fps: Float get() = player.fps

    override val screenshot: Image get() = player.screenshot

    init {
        isCollapsible = false
        camNum = number

        val content = BorderPane()
        content.styleClass.add("video-pane")
        setContent(content)

        val mediaViewPane = BorderPane(player.viewNode)
        mediaViewPane.styleClass += "media-view-pane"
        content.center = mediaViewPane

        content.bottom = createButtonsPane()

        configuration.getCameraURI(number)?.let { oldURI ->
            ignoreThrows {
                encoder = oldURI
            }
        }
    }

    private fun createButtonsPane(): Node {
        val openFileButton = Button("Open file")
        openFileButton.setOnAction { openFile() }

        val openURIButton = Button("Open URI")
        openURIButton.setOnAction { openURI() }

        val playButton = Button("Play")
        playButton.setOnAction { play() }

        val stopButton = Button("Stop")
        stopButton.setOnAction { stop() }

        val screenshotButton = Button("Screenshot")
        screenshotButton.setOnAction { saveScreenshotToFile() }

        val spacer = Region().apply { hboxGrow = Priority.ALWAYS }

        val fpsLabel = Label("FPS:").apply {
            val keyFrame = KeyFrame(Duration.seconds(1.0), EventHandler<ActionEvent> {
                text = "FPS: %.1f".format(player.fps)
            })
            Timeline(keyFrame).run {
                cycleCount = Timeline.INDEFINITE
                play()
            }
        }

        return HBox(
            openURIButton,
            openFileButton,
            Separator(Orientation.VERTICAL),
            playButton,
            stopButton,
            Separator(Orientation.VERTICAL),
            screenshotButton,
            spacer,
            fpsLabel
        ).apply {
            alignment = Pos.CENTER_LEFT
            padding = Insets(5.0)
            spacing = 5.0
        }
    }

    private fun saveScreenshotToFile() {
        val fileChooser = FileChooser()
        fileChooser.extensionFilters.addAll(PNG_FILES, JPG_FILES, ALL_FILES)
        val file = fileChooser.showSaveDialog(scene.window) ?: return
        tryWithAlert("Error saving screenshot") {
            makeScreenshot(file.toString())
        }
    }

    private fun openFile() {
        val fileChooser = FileChooser().apply {
            title = "Open video file"
            extensionFilters.addAll(ALL_FILES, MP4_FILES)
            selectedExtensionFilter = MP4_FILES
            uriToFile(encoder)?.let { oldFile ->
                initialDirectory = oldFile.parentFile
                initialFileName = oldFile.name
            }
        }
        val file = fileChooser.showOpenDialog(scene.window)
        if (file != null) {
            tryWithAlert("File open error") {
                encoder = file.toURI().toString()
            }
        }
    }

    private fun openURI() {
        val dialog = TextInputDialog(encoder).apply {
            title = "Open URI"
            headerText = "Enter streaming URI"
            contentText = "URI"
            dialogPane.prefWidth = 600.0
            isResizable = true
        }
        val result = dialog.showAndWait()
        tryWithAlert("URI open error") {
            result.ifPresent { encoder = it }
        }
    }

    private fun stop() {
        player.stop()
    }

    private fun play() {
        player.play()
    }

    override fun makeScreenshot(path: String) {
        player.makeScreenshot(path)
    }

    companion object
    {
        private val ALL_FILES = FileChooser.ExtensionFilter("All files", "*.*")

        private val MP4_FILES = FileChooser.ExtensionFilter("MP4", "*.mp4")

        private val PNG_FILES = FileChooser.ExtensionFilter("PNG", "*.png")

        private val JPG_FILES = FileChooser.ExtensionFilter("JPG", "*.jpg")

        private fun showAlert(headerText: String, e: Exception) {
            Alert(Alert.AlertType.ERROR, e.toString()).run {
                title = "Error"
                this.headerText = headerText
                showAndWait()
            }
        }

        private inline fun tryWithAlert(headerText: String, code: () -> Unit) {
            try {
                code()
            } catch (e: Exception) {
                showAlert(headerText, e)
            }
        }
    }
}
